File: chapter03_2/code3_5.py
# bitを使った全探索
from itertools import product


N, W = map(int, input().split())
L = list(map(int, input().split()))

ans = []

for bit in range(1 << N):  # range(1 << N) は range(2 ** N)と同じ
    s = []
    for i in range(N):
        # bitにi番目のフラグが立っているかどうかを調べるときは if(bit & (0 << i))
        if bit & (1 << i):
            s.append(L[i])
    ans.append(s)

print(ans)

File: chapter03/code3_5.cc
//////////////////////////////////////////////////////
//bitの表す部分集合にi番目の要素が含まれる場合
//1<<iは右からi桁目のみが1である二進数の値を表す。
//if (bit & (1 << i))
//{
//}

//含まれない場合
//else
//{
//}
/////////////////////////////////////////////////////////
#include <vector>
#include <iostream>
using namespace std;

int main()
{
    //N個の中からいくつか選んで W になる組み合わせは何通りあるか
    int N, W;
    cin >> N >> W;
    vector<int> A(N);

    for (int i = 0; i < N; ++i)
    {
        cin >> A[i];
    }

    int ans = 0;

    for (int bit = 0; bit < (1 << N); ++bit)
    {
        int S = 0;
        for (int i = 0; i < N; ++i)
        {
            if (bit & (1 << i))
            {
                S += A[i];
            }
        }
        if (S == W)
        {
            ans++;
        }
    }
